### PR TITLE
Batch access checks per version-store operation

### PR DESCRIPTION
**Pull request by snazy**
This change allows collecting all "AccessChecker.canXyz()" checks to process all of them at once.

Introduces a new interface `Authorizer` which acts as a factory to produce instances of `AccessChecker`.
`AccessChecker` gets two new methods:
* `check()` to return all failled checks.
* `checkAndThrow()`, implemented as a default method on `AccessChecker` to throw an `AccessControlException`, if `check()` returns at least one failed check.

`AbstractAccessChecker` contains implementations of all the `canXyz()` methods and collects `Check` objects describing the individual checks.

### CODE REVIEW

File: servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+              checker.canCreateReference(c.ref());
+              break;
+            case DELETE_REFERENCE:
+              checker.canDeleteReference(c.ref());
+              break;
+            case READ_ENTRIES:
+              checker.canReadEntries(c.ref());
+              break;
+            case ASSIGN_REFERENCE_TO_HASH:
+              E:
```
Comment (nastra) [inline servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java:137]: can be removed I guess?

File: servers/services/src/main/java/org/projectnessie/services/authz/Check.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+package org.projectnessie.services.authz;
+
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.versioned.NamedRef;
+
+@Value.Immutable
+public interface Check {
+  CheckType name();
```
Comment (nastra) [inline servers/services/src/main/java/org/projectnessie/services/authz/Check.java:25]: nit: type()

File: servers/services/src/main/java/org/projectnessie/services/authz/AccessChecker.java [hunk 3]
```diff
@@ -16,135 +16,129 @@
...
+ * Accept various allowance-checks and retrieve the result of all operations at once.
+ *
+ * <p>The purpose of the {@link AccessChecker} is to accept all required checks via the {@code
+ * can...()} methods and return the result of these "can do xyz" checks via {@link #check()}.
+ *
+ * <p>The checks make sure that a particular role is allowed to perform an action (such as creation,
+ * deletion) on a {@link NamedRef} (Branch/Tag). Additionally, this interface also provides checks
+ * based on a given {@link ContentKey}.
  */
 public interface AccessChecker {
```
Comment (nastra) [inline servers/services/src/main/java/org/projectnessie/services/authz/AccessChecker.java:34]: updates in this file also need to be reflected in the markdown file for authorization

File: servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+              checker.canUpdateEntity(c.ref(), c.key(), c.contentId());
+              break;
+            case DELETE_ENTITY:
+              checker.canDeleteEntity(c.ref(), c.key(), c.contentId());
+              break;
+            case VIEW_REFLOG:
+              checker.canViewRefLog();
+              break;
+            default:
+              throw new IllegalArgumentException("Unkupported: " + c);
```
Comment (nastra) [inline servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java:159]: typo

**Review by nastra — CHANGES_REQUESTED** [review]
overall LGTM, just a few small things

File: servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java (hunk 1 above)
Comment (snazy) [reply servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java:137]: Can't recall that I've intentionally added that - duh

Comment (nastra) [reply servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java:137]: the cat/dog was probably typing while you were away :)

File: servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java (hunk 4 above)
Comment (snazy) [reply servers/services/src/test/java/org/projectnessie/services/authz/AccessCheckerTest.java:159]: good eyes

**Review by nastra — APPROVED** [review]
looks great

**Pull request merged**